src/algokit/core/dispenser.py

```diff
@@ -0,0 +1,199 @@
+    *, url_suffix: str, data: dict | None = None, ci: bool, method: str = "POST"
+) -> httpx.Response | None:
+    """Generalized method to process http requests to dispenser API"""
+
+    if not ci and not is_authenticated():
+        logger.error("Please login first")
+        return None
+
+    headers = {"Authorization": f"Bearer {get_auth_token(ci=ci)}"}
+
+    # Set request arguments
+    request_args = {
+        "url": f"{AUTH_CONFIG.base_url}/{url_suffix}",
+        "headers": headers,
+        "timeout": DISPENSER_REQUEST_TIMEOUT,
+    }
+
+    if method.upper() != "GET" and data is not None:
+        request_args["json"] = data
+
+    try:
+        response: httpx.Response = getattr(httpx, method.lower())(**request_args)
+        return response
+    except Exception as err:
+        error_message = "Error processing dispenser API request"
+        logger.debug(f"{error_message}: {err}", exc_info=True)
+        raise Exception(error_message) from err
+
+
+def set_keyring_passwords(token_data: dict[str, str]) -> None:
+    decoded_id_token = jwt.decode(token_data["id_token"], algorithms=ALGORITHMS, options={"verify_signature": False})
+    user_id = decoded_id_token.get("sub")
+    data = AccountKeyringData(
+        id_token=token_data["id_token"],
+        access_token=token_data["access_token"],
+        refresh_token=token_data.get("refresh_token", ""),
+        user_id=user_id or "",
+    )
+    keyring.set_password(KEYRING_NAMESPACE, KEYRING_KEY, json.dumps(data.__dict__))
+
+
+def get_keyring_data() -> AccountKeyringData:
+    data_str = keyring.get_password(KEYRING_NAMESPACE, KEYRING_KEY)
+    if not data_str:
+        raise Exception("No keyring data found")
+    return AccountKeyringData(**json.loads(data_str))
+
+
+def is_authenticated() -> bool:
+    try:
+        data = get_keyring_data()
+    except Exception:
+        return False
+    decoded_token = jwt.decode(data.access_token, options={"verify_signature": False})
+    return bool(time.time() < decoded_token.get("exp", 0))
+
+
+def validate_id_token(id_token: str, audience: str) -> None:
+    jwks_url = f"https://{AUTH_CONFIG.domain}/.well-known/jwks.json"
+    issuer = f"https://{AUTH_CONFIG.domain}/"
+    sv = AsymmetricSignatureVerifier(jwks_url)
+    tv = TokenVerifier(signature_verifier=sv, issuer=issuer, audience=audience)
+    tv.verify(id_token)
+
+
+def refresh_token() -> None:
+    data = get_keyring_data()
+    headers = {
+        "content-type": "application/x-www-form-urlencoded",
+        "authorization": f"Bearer {data.access_token}",
+    }
+    token_data = {
+        "grant_type": "refresh_token",
+        "client_id": AUTH_CONFIG.client_ids[DispenserApiAudiences.USER],
+        "refresh_token": data.refresh_token,
+    }
+    response = httpx.post(
+        f"https://{AUTH_CONFIG.domain}/oauth/token", data=token_data, headers=headers, timeout=DISPENSER_REQUEST_TIMEOUT
+    )
+    if response.status_code != httpx.codes.OK:
```

Comment:
Should we more gracefully handle the scenario where a refresh token has expired and that's the reason we can't refresh the tokens?

Reply:
@neilcampbell should be placing a descriptive logger error to user if refresh failed